library/Fission/Storage/IPFS/Add.hs

```diff
@@ -0,0 +1,167 @@
+            |> mkCID
+            |> IPFS.Pin.add
+
+        bad ->
+          return . Left . UnexpectedOutput <| UTF8.textShow bad
+
+    (ExitFailure _, _, err) ->
+      return . Left . UnknownAddErr <| UTF8.textShow err
+
+addFile ::
+  ( MonadRIO          cfg m
+  , HasProcessContext cfg
+  , HasLogFunc        cfg
+  , Has HTTP.Manager  cfg
+  , Has IPFS.URL      cfg
+  , Has IPFS.BinPath  cfg
+  , Has IPFS.Timeout  cfg
+  )
+  => Lazy.ByteString
+  -> IPFS.Name
+  -> m (Either IPFS.Error.Add IPFS.SparseTree)
+addFile raw name =
+  IPFS.Proc.run opts raw >>= \case
+    (ExitSuccess, result, _) ->
+      IPFS.Pin.add (mkCID <| UTF8.textShow result) >>= \case
+        Left err ->
+          return . Left . UnknownAddErr <| UTF8.textShow err
+
+        Right _ ->
+          case CL.lines result of
+            [inner, outer] ->
+              let
+                sparseTree  = Directory [(Hash rootCID, fileWrapper)]
+                fileWrapper = Directory [(fileName, Content fileCID)]
+                rootCID     = CID <| UTF8.textShow outer
+                fileCID     = CID . UTF8.stripN 1 <| UTF8.textShow inner
+                fileName    = Key name
+              in
+                return <| Right sparseTree
+
+            bad ->
+              return . Left . UnexpectedOutput <| UTF8.textShow bad
+
+
+    (ExitFailure _, _, err) ->
+      return . Left . UnknownAddErr <| UTF8.textShow err
+
+    where
+      opts = [ "add"
+             , "-wq"
+             , "--stdin-name"
+             , unName name
+             ]
+
+addPath ::
+  ( RIOProc           cfg m
+  , Has IPFS.Timeout  cfg
+  , Has IPFS.BinPath  cfg
+  )
+  => FilePath
+  -> m (Either IPFS.Error.Add CID)
+addPath path = IPFS.Proc.run ["add", "-HQ", path] "" >>= pure . \case
+    (ExitSuccess, result, _) ->
+      case CL.lines result of
+        [cid] -> Right . mkCID . UTF8.stripN 1 <| UTF8.textShow cid
+        bad   -> Left . UnexpectedOutput <| UTF8.textShow bad
+
+    (ExitFailure _, _, err) ->
+      Left . UnknownAddErr <| UTF8.textShow err
+
+addDir ::
+  ( MonadRIO cfg m
+  , HasProcessContext cfg
+  , HasLogFunc cfg
+  , Has IPFS.BinPath cfg
+  , Has IPFS.Timeout cfg
+  )
+  => [Glob.Pattern]
+  -> FilePath
+  -> m (Either IPFS.Error.Add IPFS.CID)
+addDir ignored path = doesFileExist path >>= \case
+  True -> addPath path
+  False -> walkDir ignored path
+
+walkDir ::
+  ( MonadRIO cfg m
+  , HasProcessContext cfg
+  , HasLogFunc cfg
+  , Has IPFS.BinPath cfg
+  , Has IPFS.Timeout cfg
+  )
+  => [Glob.Pattern]
+  -> FilePath
+  -> m (Either IPFS.Error.Add IPFS.CID)
+walkDir ignored path = do
+  files <- listDirectory path
+  let toAdd = removeIgnored ignored files
+  results <- mapM (addDir ignored . combine path) toAdd 
+  case lefts results of
+    [] -> do
+      let names = IPFS.Name <$> toAdd
+      links <- zipWithM DAG.Link.create (rights results) names
+      case lefts links of
+        [] -> DAG.putNode Node
+              { dataBlock = "CAE="
```

Comment:
Never did find out what this mysterious data is :sob:. Still looking around. Parsing w/ the multiformats library gave me an error :thinking:

Reply:
`CAE=` == `\u0008\u0001`
https://medium.com/@php.laboratory/you-are-right-that-mysterious-u0008-u0001-came-from-protobuf-its-serialized-directory-1-from-dc3a42b59db8

This is from protobufs and `CAE=` just means "directory". I guess we can leave this hardcoded for now? It's the only type of IPLD node we need to manually generate atm. If/when we start adding other types of IPLD nodes, we could look into implementing protobufs for haskell or switching over to a different dag format

Reply:
Actually [dag-pb](https://github.com/ipld/go-ipld-deprecated) seems to be deprecated. But it's still the default when doing a recursive add. Some experimenting with `dag-cbor` didn't work with the gateway, but there's the possibility that I'm doing something incorrectly there